pulser-core/pulser/result.py

```diff
@@ -0,0 +1,133 @@
+# Copyright 2023 Pulser Development Team
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Classes to store measurement results."""
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from collections import Counter
+from dataclasses import dataclass
+
+import matplotlib.pyplot as plt
+import numpy as np
+
+from pulser.register import QubitId
+
+
+@dataclass  # type: ignore
+class Result(ABC):
+    """Base class for storing the result of a sequence run."""
+
+    atom_order: tuple[QubitId, ...]
+    meas_basis: str
+
+    @property
+    def probability_dist(self) -> dict[str, float]:
+        """Probability distribution of each bitstring."""
+        n = self._size
+        return {
+            np.binary_repr(ind, width=n): prob
+            for ind, prob in enumerate(self._weights())
+            if prob != 0
+        }
+
+    @property
+    @abstractmethod
+    def sampling_errors(self) -> dict[str, float]:
+        """The sampling error associated to each bitstring probability."""
+        pass
+
+    @property
+    def _size(self) -> int:
+        return len(self.atom_order)
+
+    @abstractmethod
+    def _weights(self) -> np.ndarray:
+        """Contains the probability of every state in an ordered array."""
+        pass
+
+    def sample(self, n_samples: int) -> Counter[str]:
```

Comment:
Don't we want to also have access to the state directly (for simulators) rather than only the samples? This would likely mean adding another (I guess abstract) method.

Reply:
That's the thing, for now we can only get the state when it is coming from Qutip (plus it is given as a `qutip.Qobj`), so I defined `get_state()` inside `QutipResult` instead.

Reply:
Ah got it. It makes sense, thanks @a-corni.